sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala

```diff
@@ -752,286 +951,173 @@ class GpuMergeAggregateIterator(
-          // second pass agg
-          tryMergeAggregatedBatches()
-
-          val rowsAfterSecondPassAgg = aggregatedBatches.asScala.foldLeft(0L) {
-            (totalRows, batch) => totalRows + batch.numRows()
-          }
-          shouldSkipThirdPassAgg =
-            rowsAfterSecondPassAgg > skipAggPassReductionRatio * rowsAfterFirstPassAgg
-        } else {
-          shouldSkipThirdPassAgg = true
-          logInfo(s"Rows after first pass aggregation $rowsAfterFirstPassAgg exceeds " +
-            s"${skipAggPassReductionRatio * 100}% of " +
-            s"localInputRowsCount ${localInputRowsCount.value}, skip the second pass agg")
-        }
-      }
+    realIter.map(_.next()).getOrElse {
 
-      if (aggregatedBatches.size() > 1) {
-        // Unable to merge to a single output, so must fall back
-        if (allowNonFullyAggregatedOutput && shouldSkipThirdPassAgg) {
-          // skip third pass agg, return the aggregated batches directly
-          logInfo(s"Rows after second pass aggregation exceeds " +
-            s"${skipAggPassReductionRatio * 100}% of " +
-            s"rows after first pass, skip the third pass agg")
-          fallbackIter = Some(new Iterator[ColumnarBatch] {
-            override def hasNext: Boolean = !aggregatedBatches.isEmpty
-
-            override def next(): ColumnarBatch = {
-              withResource(aggregatedBatches.pop()) { spillableBatch =>
-                spillableBatch.getColumnarBatch()
-              }
-            }
-          })
-        } else {
-          // fallback to sort agg, this is the third pass agg
-          fallbackIter = Some(buildSortFallbackIterator())
+      // Handle reduction-only aggregation
+      if (isReductionOnly) {
+        val batches = ArrayBuffer.apply[SpillableColumnarBatch]()
+        while (firstPassIter.hasNext) {
+          batches += firstPassIter.next()
         }
-        fallbackIter.get.next()
-      } else if (aggregatedBatches.isEmpty) {
-        if (hasReductionOnlyBatch) {
+
+        if (batches.isEmpty || batches.forall(_.numRows() == 0)) {
           hasReductionOnlyBatch = false
-          generateEmptyReductionBatch()
+          return generateEmptyReductionBatch()
         } else {
-          throw new NoSuchElementException("batches exhausted")
+          hasReductionOnlyBatch = false
+          val concat = AggregateUtils.concatenateAndMerge(batches, metrics, concatAndMergeHelper)
+          return withResource(concat) { cb =>
+            cb.getColumnarBatch()
+          }
         }
-      } else {
-        // this will be the last batch
-        hasReductionOnlyBatch = false
-        withResource(aggregatedBatches.pop()) { spillableBatch =>
-          spillableBatch.getColumnarBatch()
+      }
+
+      // Handle the case of skipping second and third pass of aggregation
+      // This only work when spark.rapids.sql.agg.skipAggPassReductionRatio < 1
+      if (!firstBatchChecked && firstPassIter.hasNext
```

Comment:
If we are doing an aggregate every time, would it be better to check each batch and skip repartitioning if the batch stayed large?

Reply:
I don't quite understand this. can you elaborate on this?

Reply:
This is just a follow on issue. Right now we check the first batch and decide for the entire task if we want to skip full aggregation or not.  I think it would probably be better if we did it for each batch individually.  We could even do it for each batch when we merge them too.

Reply:
ok, I can open a follow up issue on this. The question is, if we have already skipped agg for first batch, what should we do when second batch does not meet skip criterial?  The first batch has already been transferred to next operator by now.

Reply:
issue opened at https://github.com/NVIDIA/spark-rapids/issues/11770